src/deployconfig.jl

```diff
@@ -843,6 +843,215 @@ authentication_method(::Buildkite) = Documenter.SSH
+        deploy_branch = branch
+        deploy_repo = repo
+        is_preview = false
+        ## Deploy to deploydocs devurl kwarg
+        subfolder = devurl
+    else # build_type === :preview
+        m = match(r"refs\/pull\/(\d+)\/merge", cfg.woodpecker_ref)
+        pr_number1 = tryparse(Int, (m === nothing) ? "" : m.captures[1])
+        pr_number2 = tryparse(Int, get(ENV, "CI_COMMIT_PULL_REQUEST", nothing) === nothing ? "" : ENV["CI_COMMIT_PULL_REQUEST"])
+        # Check if both are Ints. If both are Ints, then check if they are equal, otherwise, return false
+        pr_numbers_ok = all(x -> x isa Int, [pr_number1, pr_number2]) ? (pr_number1 == pr_number2) : false
+        is_pull_request_ok = get(ENV, "CI_BUILD_EVENT", "") == "pull_request"
+        pr_ok = pr_numbers_ok == is_pull_request_ok
+        all_ok &= pr_ok
+        println(io, "- $(marker(pr_numbers_ok)) ENV[\"CI_COMMIT_REF\"] corresponds to a PR")
+        println(io, "- $(marker(is_pull_request_ok)) ENV[\"CI_BUILD_EVENT\"] matches built type: `pull_request`")
+        btype_ok = push_preview
+        all_ok &= btype_ok
+        println(io, "- $(marker(btype_ok)) `push_preview` keyword argument to deploydocs is `true`")
+        deploy_branch = branch_previews
+        deploy_repo = repo_previews
+        is_preview = true
+        ## deploydocs to previews/PR
+        subfolder = "previews/PR$(something(pr_number1, 0))"
+    end
+
+    token_ok = env_nonempty("PROJECT_ACCESS_TOKEN")
+    auth_ok = token_ok
+    all_ok &= auth_ok
+
+    if token_ok
+        println(io, "- $(marker(token_ok)) ENV[\"PROJECT_ACCESS_TOKEN\"] exists and is non-empty")
+    end
+
+    print(io, "Deploying: $(marker(all_ok))")
+    @info String(take!(io))
+    if build_type === :devbranch && !branch_ok && devbranch == "master" && cfg.woodpecker_ref == "refs/heads/main"
+        @warn """
+        Possible deploydocs() misconfiguration: main vs master. Current branch (from \$CI_COMMIT_REF) is "main". 
+        """
+    end
+
+    if all_ok
+        return DeployDecision(; all_ok=true,
+            branch=deploy_branch,
+            is_preview=is_preview,
+            repo=deploy_repo,
+            subfolder=subfolder
+        )
+    else
+        return DeployDecision(; all_ok=false)
+    end
+end
+
+authentication_method(::Woodpecker) = Documenter.HTTPS
+function authenticated_repo_url(cfg::Woodpecker)
+    # `cfg.woodpecker_forge_url` should be just the root of the URL e.g. github.com, gitlab.com, codeberg.org
+    # otherwise, it will be equal to `cfg.woodpecker_repo_link`
+    # If so, we just split the `http(s)://` from the string we want 
+    # e.g. `https://github.com/JuliaDocs/Documenter.jl` to `github.com/JuliaDocs/Documenter.jl`.
+    if haskey(ENV, "FORGE_URL")
+        return "https://$(ENV["CI_REPO_OWNER"]):$(ENV["PROJECT_ACCESS_TOKEN"])@$(ENV["FORGE_URL"])/$(cfg.woodpecker_repo).git"
```

Comment:
I wonder if Documenter might leak this URL to the logs. For GitHub actions, which also uses an authenticated URL like this it isn't much of a problem since there you use a temporary token.

Reply:
The token will be shown as `****` in the logs so the token isn't leaked. I am not sure what you mean leaking of the URL. you mean that Documenter may show `https://JuliaDocs:myprojectaccesstoken@github.com/JuliaDocs/Documenter.jl`? It will only show as `https://****:****@github.com/****/Documenter.jl`.

Reply:
Although, unless they `echo`-ed their secrets to a file, then I'm going to question their life choices

Reply:
You can see a successful log here https://ci.codeberg.org/SynBioJulia/SBOL3.jl/build/6/3. There is nothing much you can't see, even the url passed from the `authenticated_repo_url` function is not showing.

Reply:
I wonder if it would be possible to use [`git-credential-store`](https://git-scm.com/docs/git-credential-store) for this? I don't have experience with it, but it looks like it would then be similar to how the handle the SSH keys.

Reply:
I don't think having to use the git credential store is necessary. git credential store just does what with the URL we have to authenticate stored as a text file. We are just adding an extra step that I don't think is needed.

Reply:
Although I have an idea. We can use https://github.com/str4d/rage or https://github.com/FiloSottile/age for this if you want to push with the git credential store. it is an extra step. you juts need to encrypt the credential store file with the corresponding URLs, add it to your project repo and then deencrypt it inside the container. It is many steps but I think that is what you want to avoid leaking of information?

Reply:
Anyway seems we can use SSH. I guess I am just overthinking a lot. See https://github.com/JuliaDocs/Documenter.jl/pull/1880/commits/c7c5bf1d0ba549a4d05ab32b77465daa568a2bba and https://ci.codeberg.org/SynBioJulia/SBOL3.jl/build/12 that now uses `DOCUMENTER_KEY`. I will add the documentation for creating the keys.

Reply:
Hmm the logs for that build say

```
│ - ✔ ENV["PROJECT_ACCESS_TOKEN"] exists and is non-empty
```

? Did it actually use the SSH key?

Reply:
Also, just for future reference, here is where Documenter would happily leak the secret if it were to run into an error

https://github.com/JuliaDocs/Documenter.jl/blob/396ee24391c081792e2f3cc32bb47699d7ddc3dd/src/Documenter.jl#L631-L642

While the censoring by the CI would generally catch this, I think ideally we should avoid this altogether. But that can be a subject of a follow-up PR (#1889). Just for the record though, I assume we'd have Documenter write the credentials file on every `deplodocs` run, just like we do with the SSH key.